### PR TITLE
catalog: refactor pg_type view definition

### PR DESCRIPTION
**Pull request by sploiselle**
I realized `list` types weren't included in the `pg_type` views, and when I added them the catalog went from 1s to open to 24s.

```diff
diff --git a/src/coord/src/catalog/builtin.rs b/src/coord/src/catalog/builtin.rs
index 672f82214..dfba0d804 100644
--- a/src/coord/src/catalog/builtin.rs
+++ b/src/coord/src/catalog/builtin.rs
@@ -1027,6 +1027,7 @@ pub const PG_TYPE: BuiltinView = BuiltinView {
     CASE
         WHEN EXISTS (SELECT 1 FROM mz_catalog.mz_array_types WHERE type_id = mz_types.id) THEN 'a'
         WHEN EXISTS (SELECT 1 FROM mz_catalog.mz_base_types WHERE type_id = mz_types.id) THEN 'b'
+        WHEN EXISTS (SELECT 1 FROM mz_catalog.mz_list_types WHERE type_id = mz_types.id) THEN 'l'
         WHEN EXISTS (SELECT 1 FROM mz_catalog.mz_map_types WHERE type_id = mz_types.id) THEN 'm'
     END AS typtype,
     0::pg_catalog.oid AS typrelid,
```

I played science dog with this for a while to figure out what was causing the blowup and the only difference between the raw plans is

```
Map if exists(%3) then {"a"} els {if exists(%4) then {"b"} els {if exists(%5) then {"m"} els {null}}}, i32tooid(0), null, coalesce(select(%8), i32tooid(0)), null, false, i32tooid(0)
```
becomes
```
Map if exists(%3) then {"a"} els {if exists(%4) then {"b"} els {if exists(%5) then {"l"} els {if exists(%6) then {"m"} els {null}}}}, i32tooid(0), null, coalesce(select(%9), i32tooid(0)), null, false, i32tooid(0)
```

so it seems like adding these layers of `if els` contributes to the slowdown, so I instead rewrote this view definition to simply perform a join. This rewrite is ~the same as the old version, where opening the catalog takes ~1s.

idk if this speaks to an underlying issue with our ability to handle `CASE...WHEN...` clauses, but glad to spend some time looking into it more deeply if it seems worthwhile.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5368)

### CODE REVIEW

**Review by justinj — COMMENTED** [review]
The slowdown is interesting, I wonder if there's some bad patterns getting triggered in the decorrelation of it. Probably worth filing an issue, seems like a possible opportunity for some improvements! Change :lgtm:

> *[Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5368#-:-MRQ8UmGCZwj_zqNYkAu:b-s2sawc)* status: 0 of 1 files reviewed, 1 unresolved discussion (waiting on @benesch and @sploiselle)

___
*[src/coord/src/catalog/builtin.rs, line 1046 at r1](https://reviewable.io/reviews/materializeinc/materialize/5368#-MRQ9I4nEH3Bm98Gv-0X:-MRQ9I4nEH3Bm98Gv-0Y:bw39prs) ([raw file](https://github.com/materializeinc/materialize/blob/4fd285c68424014eae7ba603094ed29d4a6cab3e/src/coord/src/catalog/builtin.rs#L1046)):*
> ```Rust
> JOIN (
>     SELECT type_id, 'a' AS typtype FROM mz_catalog.mz_array_types
>         UNION SELECT type_id, 'b' FROM mz_catalog.mz_base_types
> ```

not sure if we can automatically demote this to a `union all` but if not, there would be some (minor) efficiency gains from doing so (not sure if the scale of these views merits that)

File: src/coord/src/catalog/builtin.rs [hunk 1]
```diff
@@ -1044,7 +1040,14 @@ pub const PG_TYPE: BuiltinView = BuiltinView {
...
 FROM mz_catalog.mz_types
-JOIN mz_catalog.mz_schemas ON mz_schemas.id = mz_types.schema_id",
+JOIN mz_catalog.mz_schemas ON mz_schemas.id = mz_types.schema_id
+JOIN (
+    SELECT type_id, 'a' AS typtype FROM mz_catalog.mz_array_types
+        UNION SELECT type_id, 'b' FROM mz_catalog.mz_base_types
+        UNION SELECT type_id, 'l' FROM mz_catalog.mz_list_types
+        UNION SELECT type_id, 'm' FROM mz_catalog.mz_map_types
+    ) AS t
+    ON mz_types.id = t.type_id",
```
Comment (benesch) [inline src/coord/src/catalog/builtin.rs:1050]: I think this line and the line above it want to be unindented to the level of `JOIN`.

**Review by benesch — APPROVED** [review]
Great find! I bet there are some other places where we're tripping over decorrelation issues like this.

One thing I am confused about: why is this impacting catalog _startup_? I wouldn't expect any of these plans to get run until someone actually queries `pg_type`?

**Comment by sploiselle** [comment]
@benesch I can spend some time poking at this but my best guess is that it gets stuck in the optimizer––not that I can substantiate that, it's just the only place where anything really happens on startup.

**Comment by sploiselle** [comment]
@justinj good call; I noticed the unnecessary distincts but didn't figure it mattered much. didn't realize removing them would be so trivial, so definitely worth doing.

File: src/coord/src/catalog/builtin.rs [hunk 2]
```diff
@@ -1024,27 +1024,35 @@ pub const PG_TYPE: BuiltinView = BuiltinView {
...
-    CASE
-        WHEN EXISTS (SELECT 1 FROM mz_catalog.mz_array_types WHERE type_id = mz_types.id) THEN 'a'
-        WHEN EXISTS (SELECT 1 FROM mz_catalog.mz_base_types WHERE type_id = mz_types.id) THEN 'b'
-        WHEN EXISTS (SELECT 1 FROM mz_catalog.mz_map_types WHERE type_id = mz_types.id) THEN 'm'
-    END AS typtype,
+    typtype,
     0::pg_catalog.oid AS typrelid,
     NULL::pg_catalog.oid AS typelem,
-    coalesce(
+    COALESCE(
```
Comment (benesch) [inline src/coord/src/catalog/builtin.rs:1030]: Don't bother going back through CI for this, but per PostgreSQL style (https://github.com/MaterializeInc/materialize/blob/main/doc/developer/style.md#sql-style) this is supposed to be lowercase!

**Review by benesch — APPROVED** [review]
LGTM!

File: src/coord/src/catalog/builtin.rs (hunk 2 above)
Comment (sploiselle) [reply src/coord/src/catalog/builtin.rs:1030]: face palm––a bug in sqlfum.pt

Comment (sploiselle) [reply src/coord/src/catalog/builtin.rs:1030]: @mjibson in case you're interested!

Comment (mjibson) [reply src/coord/src/catalog/builtin.rs:1030]: Interesting! Maybe one day...

**Pull request merged**